jerry-core/parser/js/opcodes-dumper.cpp

```diff
@@ -22,7 +22,108 @@
 /**
  * Register allocator's counter
  */
-static vm_idx_t jsp_reg_next;
+// static vm_idx_t jsp_reg_next;
+
+
+/**
+ * Bit field, indicating which of registers are free or occupied. Each bit of the buffer corresponds to one
+ * general register in the following way:
+ *
+ * free_regs_mask[0] bit 0   -   VM_REG_GENERAL_FIRST
+ * free_regs_mask[0] bit 1   -   VM_REG_GENERAL_FIRST + 1
+ * ....
+ * free_regs_mask[0] bit 31  -   VM_REG_GENERAL_FIRST + 31
+ * free_regs_mask[1] bit 0   -   VM_REG_GENERAL_FIRST + 32
+ * ....
+ *
+ * If bit is set to 1, the corresponding register is occupied, otherwise it could be used.
+ */
+static uint32_t free_regs_mask[4];
+
+JERRY_STATIC_ASSERT (sizeof (free_regs_mask) * JERRY_BITSINBYTE >= VM_REG_GENERAL_LAST - VM_REG_GENERAL_FIRST);
+
+#define REGS_IN_DWORD 32
+#define GENERALS_REGS_COUNT (VM_REG_GENERAL_LAST - VM_REG_GENERAL_FIRST + 1)
+
+/**
+ * Status of a general register: free/occupied
+ */
+typedef enum
+{
+  REG_FREE,
+  REG_OCCUPIED
+} reg_status_t;
+
+/**
+ * Sets status of a general register in free_regs_mask bit field
+ *
+ */
+static void
+set_reg_status (vm_idx_t reg, /**< register number, register out of the range VM_REG_GENERAL_FIRST -
+                            * VM_REG_GENERAL_LAST are ignored */
+                reg_status_t reg_status) /**< status of the register */
+{
+  if (!(reg >= VM_REG_GENERAL_FIRST && reg <= VM_REG_GENERAL_LAST))
+  {
+    return;
+  }
+
+  uint32_t *free_regs_mask_p;
+
+  int index = (reg - VM_REG_GENERAL_FIRST) / REGS_IN_DWORD;
+  reg = (vm_idx_t) (reg - VM_REG_GENERAL_FIRST - REGS_IN_DWORD * index);
+  free_regs_mask_p = &free_regs_mask[index];
+
+  uint32_t shifted_value = (uint32_t) 1 << reg;
+  if (reg_status)
+  {
+    (*free_regs_mask_p) |= shifted_value;
+  }
+  else
+  {
+    (*free_regs_mask_p) &= ~shifted_value;
+  }
+} /* set_reg_status */
+
+/**
+ * Find a free register in free_regs_mask bit field
+ */
+static vm_idx_t
+find_free_reg (void)
```

Comment:
Based on the functionality a better name would be `get_free_reg`. As this method not just finds the free register, but marks it as used.